### ISSUE TITLE
Duplicated record in location table when running opensips cluster

### ISSUE THREAD

**Issue by gitforxh** [OpenSIPS/opensips#1365]
Running multiple opensips instances as a cluster with version opensips 2.4.0-1.

usrloc settings are as following:

modparam("usrloc", "sql_write_mode", "write-through")
modparam("usrloc", "restart_persistency", "load-from-sql")
modparam("usrloc", "cluster_mode", "full-sharing")
modparam("usrloc", "skip_replicated_db_ops", 1)
modparam("usrloc", "location_cluster", 50)
...
modparam("registrar", "max_contacts", 1)
...

Here's the step to produce duplicated entries in location table:

1. Register a softphone via opensips instance 1
2. Kill the softphone (without doing unregister)
3. Restart the softphone and register it via opensips instance 2

then duplicated records like below are created:

mysql> SELECT * FROM location WHERE username='HP-0004F2675C22-2'\G;
*************************** 1. row ***************************
   contact_id: 1894256224518689811
     username: HP-0004F2675C22-2
       domain: NULL
      contact: sip:HP-0004F2675C22-2@10.68.49.4:46950;rinstance=ccd896a04f1a4a07;transport=UDP
     received: sip:10.68.49.4:46950
         path: NULL
      expires: 1526617182
            q: -1.00
       callid: NDEyYjYwZTRkYTY4NDNhOGUxODA2YjkxNjgzOGNjOTY.
         cseq: 2
last_modified: 2018-05-18 03:19:42
        flags: 0
       cflags: NAT
   user_agent: Z 3.3.21933 r21903
       socket: udp:69.168.212.27:5060
      methods: 5951
 sip_instance: NULL
     kv_store: NULL
         attr: NULL
*************************** 2. row ***************************
   contact_id: 1894256224518680150
     username: HP-0004F2675C22-2
       domain: NULL
      contact: sip:HP-0004F2675C22-2@10.68.49.4:56262;rinstance=9a7c15cd416fd546;transport=UDP
     received: sip:10.68.49.4:56262
         path: NULL
      expires: 1526617145
            q: -1.00
       callid: N2MyMjdjYjJmMzY1YTViOThmMWFlMmY0NDUwZjMyMTM.
         cseq: 2
last_modified: 2018-05-18 03:19:05
        flags: 0
       cflags: NAT
   user_agent: Z 3.3.21933 r21903
       socket: udp:69.168.212.28:5060
      methods: 5951
 sip_instance: NULL
     kv_store: NULL
         attr: NULL
2 rows in set (0.19 sec)

**Comment by gitforxh** [comment]
Have tried calling remove() function before save() like below, still can't remove duplicated records.

                remove("location","$fu");
                if (!save("location", "c1p1")) {
                        sl_reply_error();
                        exit;
                }

**Comment by bogdan-iancu** [comment]
@gitforxh , of course there are 2 records (and they are not duplicated) as there 2 different registrations - see the different callid in the records. You actually have 2 contacts registered for the same user. To control multiple contacts per user, see the `cNN` and `f` flags from [save()](http://www.opensips.org/html/docs/modules/2.4.x/registrar.html#func_save)

**Comment by gitforxh** [comment]
But note that I have max_contacts set to 1, And I called remove() to remove the AOR before calling save() and I'm passing 'c1' to save. Even by doing all these, I'm still getting duplicated records.

...
modparam("registrar", "max_contacts", 1)
...
            remove("location","$fu");
            if (!save("location", "c1p1")) {
                    sl_reply_error();
                    exit;
            }

**Comment by bogdan-iancu** [comment]
oh, sorry, I missed that.. @liviuchircu will try to reproduce and test

**Comment by liviuchircu** [comment]
@gitforxh Here is the only way in which I managed to reproduce your issue:

* Register a softphone via opensips instance 1
* **restart the whole opensips cluster**; due to `skip_replicated_db_ops`, only the original registrar will hold the contact now
* Kill the softphone (without doing unregister)
* Restart the softphone and register it via opensips instance 2
* the original registrar has 2 contacts; all other opensips instances have 1 contact

All other attempts in my case ended with a 503:

```
U 2018/05/22 16:06:18.908922 10.0.0.10:43967 -> 10.0.0.180:5060
REGISTER sip:full-sharing-cluster;transport=UDP SIP/2.0.
Via: SIP/2.0/UDP 10.0.0.10:43967;branch=z9hG4bK-d8754z-888279a3d3330a27-1---d8754z-.
Max-Forwards: 70.
Contact: <sip:liviu@10.0.0.10:43967;rinstance=fd2ff2e83c87a523;transport=UDP>.
To: <sip:liviu@full-sharing-cluster;transport=UDP>.
From: <sip:liviu@full-sharing-cluster;transport=UDP>;tag=cce5b22b.
Call-ID: MjVmYzg3ZDY4Yzc0MWY4YzJkMjYyZDI3NDc1YzQ5YjY..
CSeq: 1 REGISTER.
Expires: 3600.
Allow: INVITE, ACK, CANCEL, BYE, NOTIFY, REFER, MESSAGE, OPTIONS, INFO, SUBSCRIBE.
Supported: replaces, norefersub, extended-refer, timer, X-cisco-serviceuri.
User-Agent: Z 3.3.25608 r25552.
Allow-Events: presence, kpml.
Content-Length: 0.
.


U 2018/05/22 16:06:18.909924 10.0.0.180:5060 -> 10.0.0.10:43967
SIP/2.0 503 Service Unavailable.
Via: SIP/2.0/UDP 10.0.0.10:43967;branch=z9hG4bK-d8754z-888279a3d3330a27-1---d8754z-.
To: <sip:liviu@full-sharing-cluster;transport=UDP>;tag=0abf3ca1a25db3087a10f3db41152dae.0c97.
From: <sip:liviu@full-sharing-cluster;transport=UDP>;tag=cce5b22b.
Call-ID: MjVmYzg3ZDY4Yzc0MWY4YzJkMjYyZDI3NDc1YzQ5YjY..
CSeq: 1 REGISTER.
Contact: <sip:liviu@10.0.0.10:62132;rinstance=27868927d597ad70;transport=UDP>;expires=2731.
P-Registrar-Error: Too many registered contacts.
Server: OpenSIPS (3.0.0-dev (x86_64/linux)).
Content-Length: 0.
.
```

Does this sound good? If yes, I suggest you try the following:
* comment `skip_replicated_db_ops`. It is outdated in `2.4`, and I forgot it there. I'll remove it after this ticket. The databases are meant to be individual to each node, we have no reason to skip inserts to them.
* as per the RFC, the AoR is located inside the To header field, so you should try: `remove("location", "$fu");`. It fully cleans up the AoR in my case, across the entire cluster.

To sum up: yes, this looks like a bug, but only because of `skip_replicated_db_ops`, which should no longer be there in the first place.

**Comment by gitforxh** [comment]
@liviuchircu Thanks for the investigation.

I've tried commenting out skip_replicated_db_ops. After that, there's following error msg printed on each node of the cluster whenever I register a phone.

May 23 01:38:04 [17629] CRITICAL:db_mysql:wrapper_single_mysql_stmt_execute: driver error (1062): Duplicate entry 'HP-00156502E463-2-sip.nf-stg.fonality.com-sip:HP-00156502E463-2@' for key 'account_contact_idx'
May 23 01:38:04 [17629] ERROR:usrloc:db_insert_ucontact: inserting contact in db failed
May 23 01:38:04 [17629] ERROR:usrloc:wb_timer: inserting contact into database failed

**Comment by gitforxh** [comment]
However, I can confirm that, by commenting out skip_replicated_db_ops, the duplication in mysql is not happening anymore.

Btw, what did you mean by 'The databases are meant to be individual to each node' ? Does it mean each node need to use its own db? Can we still share mysql db among all the nodes in the cluster?

**Comment by liviuchircu** [comment]
Indeed, @gitforxh , each node must point to an individual "location" database/table, local to its box.  If you have a central DB machine and want to keep it this way, you can, of course, duplicate a separate "opensips" database for each node and specify it via the `db_url` of `usrloc`.  The `full-sharing` user location cluster is not built to work with a shared database, hence the errors you are experiencing -- if you want it to work via a shared database, you may switch to `full-sharing-cachedb` clustering.

Please let me know if you run into any more errors after updating your setup.  I will make sure all of these concepts are well documented, so the transition to `2.4` clustering is as smooth as possible.

**Comment by gitforxh** [comment]
We're actually switching to use the federation-cachedb mode now due to the NAT restriction. Does this mode support a sharing a single location table in mysql? The reason we still need the location table is because we have one other service need to get registration status of phones and it can only read from a mysql table.

**Comment by liviuchircu** [comment]
Regarding the purpose `location` table, it is very similar between "federated" and "full-sharing" modes: it's just an alternative means of ensuring restart persistency.  If your service is for-ever hardcoded to use MySQL (keep in mind you can also fetch the user location via HTTP + `mi_json` or HTTP + `mi_xmlrpc`), then you have two options:

* use two databases. Although this duplicates the data set, it will never lead to any conflicts / error states caused by a shared table.
* use a shared database, together with `skip_replicated_db_ops = 1`.  This is a feature carried from `2.3`, and I have not put much testing work into it so far (and don't plan to).  In theory, it should be fine. 
On one side, replicated contacts will not cause duplicate key errors, and contacts which expire in parallel will not lead to any SQL errors (one DELETE query works, other one is just a NOP).  On the other side, if you use a VIP with `write-back` SQL writes, you can easily trigger conflicting inserts each time you flip the VIP to the other box.  Let me know if you run into any issues with this setup, and I will gladly help -- maybe even push a fix or two.

**Comment by gitforxh** [comment]
We'll probably go with option 2 by using a shared db, but keep `skip_replicated_db_ops = 0` as according to my previous test this has successfully fixed the duplication in db. We'll just ignore all the conflicting inserts errors since it seems not causing any problem in our testing.

Thanks so much for all the help!
